### PR TITLE
chore: preparations for bolt12 support

### PR DESCRIPTION
**Pull request by joschisan**

### CODE REVIEW

File: gateway/ln-gateway/src/db.rs [hunk 1]
```diff
@@ -154,12 +154,19 @@ async fn migrate_to_v1(dbtx: &mut DatabaseTransaction<'_>) -> Result<(), anyhow:
 }
 
 #[derive(Debug, Encodable, Decodable)]
-pub struct CreateInvoicePayloadKey(pub [u8; 32]);
+pub struct RegisteredIncomingContractKey(pub [u8; 32]);
+
+#[derive(Debug, Encodable, Decodable)]
+pub struct RegisteredIncomingContract {
+    pub federation_id: FederationId,
+    pub incoming_amount: u64,
```
Comment (m1sterc001guy) [inline gateway/ln-gateway/src/db.rs:162]: `IncomingContract` has the amount in the `Commitment`, is this just for convenience?

File: modules/fedimint-lnv2-client/src/send_sm.rs [hunk 2]
```diff
@@ -43,10 +43,23 @@ pub struct SendSMCommon {
     pub funding_txid: TransactionId,
     pub gateway_api: SafeUrl,
     pub contract: OutgoingContract,
-    pub invoice: Bolt11Invoice,
+    pub invoice: Invoice,
     pub refund_keypair: KeyPair,
 }
 
+#[derive(Debug, Clone, Eq, PartialEq, Hash, Decodable, Encodable)]
+pub enum Invoice {
```
Comment (m1sterc001guy) [inline modules/fedimint-lnv2-client/src/send_sm.rs:51]: Maybe this should go in the common crate?

File: gateway/ln-gateway/src/db.rs (hunk 1 above)
Comment (m1sterc001guy) [reply gateway/ln-gateway/src/db.rs:162]: Or does this value include fees?

Comment (joschisan) [reply gateway/ln-gateway/src/db.rs:162]: The two amounts differ by the fee. incoming_amount is lower and is the amount in the invoice.

**Review by m1sterc001guy — APPROVED** [review]

File: modules/fedimint-lnv2-client/src/send_sm.rs (hunk 2 above)
Comment (joschisan) [reply modules/fedimint-lnv2-client/src/send_sm.rs:51]: I would not share this for cleaner separation.

File: gateway/ln-gateway/src/db.rs (hunk 1 above)
Comment (m1sterc001guy) [reply gateway/ln-gateway/src/db.rs:162]: Can address in the future but a different name would help IMO

**Comment by joschisan** [comment]
Please don't merge yet I am still thinking about it.

File: gateway/ln-gateway/src/db.rs (hunk 1 above)
Comment (joschisan) [reply gateway/ln-gateway/src/db.rs:162]: incoming_htlc_amount?

File: modules/fedimint-lnv2-client/src/send_sm.rs (hunk 2 above)
Comment (m1sterc001guy) [reply modules/fedimint-lnv2-client/src/send_sm.rs:54]: Why is this cleaner? I see repeated code here and in the gateway client

**Review by m1sterc001guy — APPROVED** [review]
LGTM, only comment is that it looks like to me that the `Invoice` enum could be shared between the gateway client and the user client

**Comment by Kodylow** [comment]
Note from call: moving things into enums like this is good practice for avoiding state machine migrations, @Kodylow open an issue for this

**Pull request merged**